### PR TITLE
feat(core): add support for pnpm lockfile v9

### LINKED ISSUE #22850: Production built app's package.json dependencies are empty with pnpm@9
### Current Behavior

When using pnpm 9.0.1 and creating a production build of an app, the generated `package.json` and `pnpm-lock.yaml` files contain no dependencies.

This issue does not occur with pnpm@8.15.5.

_**Update**: I've tested with the the new pnpm 9.0.2 and 9.0.5 releases and this issue still exists._

### Expected Behavior

When using pnpm 9.0.1 and creating a production build of an app, the generated `package.json` and `pnpm-lock.yaml` files should contain the application's dependencies.

### GitHub Repo

https://github.com/crbanman/nx-pnpm-9-issue

### Steps to Reproduce

1. Install pnpm@9.0.1
2. `pnpm install`
3. `pnpm nx run my-new-app:build`
4. Check `dist/my-new-app/package.json` and `dist/my-new-app/pnpm-lock.yaml` to see that their missing any dependencies

### Nx Report

```shell
Node   : 20.11.1
OS     : darwin-arm64
pnpm   : 9.0.1

nx (global)        : 18.0.5
nx                 : 18.3.0
@nx/js             : 18.3.0
@nx/jest           : 18.3.0
@nx/eslint         : 18.3.0
@nx/workspace      : 18.3.0
@nx/devkit         : 18.3.0
@nx/esbuild        : 18.3.0
@nx/eslint-plugin  : 18.3.0
@nx/node           : 18.3.0
@nrwl/tao          : 18.3.0
typescript         : 5.4.5
---------------------------------------
Registered Plugins:
@nx/eslint/plugin
@nx/jest/plugin
```

### Failure Logs

_No response_

### Package Manager Version

_No response_

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

_No response_

### LINKED ISSUE #23256: @nx/esbuild:esbuild with thirdParty: false regression from 18.3.4 to 19.0.x with pnpm
### Current Behavior

@nx/esbuild:esbuild with thirdParty: false and generatePackageJson: true removes to much while filtering pnpm-lock.yaml, so pnpm install fails when using the lock file.
[pnpm-lock.nx-19.0.1.txt](https://github.com/nrwl/nx/files/15254936/pnpm-lock.nx-19.0.1.txt)

### Expected Behavior

pnpm install works with filtered lock file, as for 18.3.4
[pnpm-lock.nx-18.3.4.txt](https://github.com/nrwl/nx/files/15254939/pnpm-lock.nx-18.3.4.txt)

### GitHub Repo

_No response_

### Steps to Reproduce

1. use esbuild to bundle with  thirdParty: false and generatePackageJson: true with nx 18.3.4
2. use esbuild to bundle with  thirdParty: false and generatePackageJson: true with nx 19.0.1
3. observe difference in generated pnpm-lock.yaml

### Nx Report

```shell
Node   : 20.12.2
OS     : win32-x64
pnpm   : 8.15.8

nx                 : 19.0.1
@nx/js             : 19.0.1
@nx/jest           : 19.0.1
@nx/linter         : 19.0.1
@nx/eslint         : 19.0.1
@nx/workspace      : 19.0.1
@nx/angular        : 19.0.1
@nx/devkit         : 19.0.1
@nx/esbuild        : 19.0.1
@nx/eslint-plugin  : 19.0.1
@nrwl/tao          : 19.0.1
@nx/web            : 19.0.1
@nx/webpack        : 19.0.1
typescript         : 5.4.5
---------------------------------------
Registered Plugins:
@nx/eslint/plugin
@nx/jest/plugin
---------------------------------------
Community plugins:
angular-calendar : 0.31.1
```

### Failure Logs

```shell
ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY  Broken lockfile: no entry for 'emoji-regex@8.0.0' in pnpm-lock.yaml

Below is items removed from working pnpm-lock.yaml for 18.3.4 attached above:

  /ansi-regex@5.0.1:
    resolution: {integrity: sha512-quJQXlTSUGL2LH9SUXo8VwsY4soanhgo6LNSm84E1LBcE8s3O0wpdiRzyR9z/ZZJMlMWv37qOOb9pdJlMUEKFQ==}
    engines: {node: '>=8'}
    dev: false

  /ansi-styles@4.3.0:
    resolution: {integrity: sha512-zbB9rCJAT1rbjiVDb2hqKFHNYLxgtk8NURxZ3IZwD3F6NtxbXZQCnnSi1Lkx+IDohdPlFp222wVALIheZJQSEg==}
    engines: {node: '>=8'}
    dependencies:
      color-convert: 2.0.1
    dev: false

  /color-convert@2.0.1:
    resolution: {integrity: sha512-RRECPsj7iu/xb5oKYcsFHSppFNnsj/52OVTRKb4zP5onXwVF3zVmmToNcOfGC+CRDpfK/U584fMg38ZHCaElKQ==}
    engines: {node: '>=7.0.0'}
    dependencies:
      color-name: 1.1.4
    dev: false

  /color-name@1.1.4:
    resolution: {integrity: sha512-dOy+3AuW3a2wNbZHIuMZpTcgjGuLU/uBL/ubcZF9OXbDo8ff4O8yVp5Bf0efS8uEoYo5q4Fx7dY9OgQGXgAsQA==}
    dev: false

  /emoji-regex@8.0.0:
    resolution: {integrity: sha512-MSjYzcWNOA0ewAHpz0MxpYFvwg6yjy1NG3xteoqz644VCo/RPgnr1/GGt+ic3iJTzQ8Eu3TdM14SawnVUmGE6A==}
    dev: false

  /is-fullwidth-code-point@3.0.0:
    resolution: {integrity: sha512-zymm5+u+sCsSWyD9qNaejV3DFvhCKclKdizYaJUuHA83RLjb7nSuGnddCHGv0hk+KY7BMAlsWeK4Ueg6EV6XQg==}
    engines: {node: '>=8'}
    dev: false
```

### Package Manager Version

_No response_

### Operating System

- [ ] macOS
- [ ] Linux
- [X] Windows
- [ ] Other (Please specify)

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by meeroslav**
- [X] Fix parenthesis separator detection
- [x] Fix leading dash detection
- [x] Migrate existing pnpm normalizer to latest code
- [X] Add unit tests for v9
- [X] Dogfooding Pnpm v9 to Nx repo and agents

Fixes regression with alias packages introduced via https://github.com/nrwl/nx/pull/23017

## Benchmarks

PNPM v9 Branch (migrated to branch's code)
```
  Time (mean ± σ):      3.526 s ±  0.081 s    [User: 0.717 s, System: 0.948 s]
  Range (min … max):    3.390 s …  3.714 s    20 runs
```
Master (running nx 19.1.0-beta.3)
```
  Time (mean ± σ):     11.160 s ±  0.112 s    [User: 0.799 s, System: 0.979 s]
  Range (min … max):   10.955 s … 11.379 s    20 runs
```

## Related Issue(s)

Fixes #22850
Fixes #23256

### CODE REVIEW

**Comment by mattfysh** [comment]
great work @meeroslav ! would it also be useful to raise an error if the `lockfileVersion` encountered is outside the expected range?

**Comment by mattfysh** [comment]
@meeroslav please don't blindly accept lockfiles whose versions are higher than 9, this is how the release of pnpm v9 caused all sorts of issues that didn't immediately map back to lockfile format support

please, if the `lockfileVersion > 9` then throw an error to indicate that nx does not support that lockfile - this will save a lot of  headaches in future

**Comment by meeroslav** [comment]
Thank you @mattfysh, that's a good feedback.

File: packages/nx/src/plugins/js/lock-file/pnpm-parser.ts [hunk 1]
```diff
@@ -326,81 +400,173 @@ export function stringifyPnpmLockfile(
...
     // so we can safely set all to prod
     matchedKeys.forEach(([key, snapshot]) => {
-      snapshot.dev = false;
-      result[key] = snapshot;
+      if (lockfileVersion >= 9) {
+        delete snapshot['dev'];
+        result[key] = snapshot;
+      } else {
+        snapshot['dev'] = false; // all dependencies are prod
+        if (lockfileVersion < 9) {
```
Comment (meeroslav) [inline packages/nx/src/plugins/js/lock-file/pnpm-parser.ts:446]: remove this

**Review by FrozenPandaz — APPROVED** [review]

**Pull request merged**